src/CameraOnvif.cpp

```diff
@@ -0,0 +1,369 @@
+#include "CameraOnvif.hpp"
+#include "SOAPException.hpp"
+
+#include <gsoap/plugin/wsseapi.h>
+#include <gsoap/plugin/wsddapi.h>
+#include "onvif/soapDeviceBindingProxy.h"
+#include "onvif/soapImagingBindingProxy.h"
+#include "onvif/soapMediaBindingProxy.h"
+#include "onvif/soapPTZBindingProxy.h"
+#include "onvif/soapPullPointSubscriptionBindingProxy.h"
+#include "onvif/soapRemoteDiscoveryBindingProxy.h"
+#include "wsdd.nsmap"
+
+#include <iostream>
+#include <sstream>
+
+using namespace std;
+using namespace camera_onvif;
+using namespace base;
+
+struct CameraOnvif::Private {
+    struct soap *soap = soap_new1(SOAP_XML_STRICT | SOAP_XML_CANONICAL | SOAP_C_UTFSTRING);
+    MediaBindingProxy *proxy_media = nullptr;
+    ImagingBindingProxy *proxy_image = nullptr;
+    _trt__GetVideoEncoderConfigurationOptionsResponse video_options;
+    Private() = default;
+    ~Private(){
+        soap_destroy(soap);
+        soap_end(soap);
+        soap_free(soap);
+
+        delete proxy_image;
+        delete proxy_media;
+    }
+};
+
+CameraOnvif::CameraOnvif(const string& user, const string& pass, const string& ip) :
+ m_user(user), m_pass(pass), m_uri("http://" + ip + "/onvif/device_service"),
```

Comment:
use the following layout:

~~~
CameraOnvif( .... )
    : m_user(user), m_pass(pass)
    , m_uri(....), ...
    , ..... {
}
~~~